include/kllvm/codegen/Decision.h

```diff
@@ -191,10 +198,11 @@ class MakePatternNode : public DecisionNode {
   }
   virtual void preprocess(std::unordered_set<LeafNode *> &leaves) {
     if (preprocessed) return;
-    child->choiceAncestors.insert(choiceAncestors.begin(), choiceAncestors.end());
     child->preprocess(leaves);
     child->predecessors.insert(this);
+    child->predecessors2.insert(this);
```

Comment:
Is there a reason to compute these incrementally instead of copying the set before sorting?

Reply:
Well, for one thing, predecessors2 is the predecessors of the dag that results from removing the outgoing edges of the failure node, so if we we used the same set, it would require more complicated logic. But also, since these sets will both become empty after `topologicalSort` is called, the difference in memory cost is at most linear in the number of nodes in the graph, and thus is relatively minor compared to the preexisting cost of representing the entire decision tree in memory. I'd rather write simpler logic now and worry about memory usage if we have examples where it is a problem. And it's not a matter of copying just one set prior to sorting because there is one set for each node. We could probably maintain a map from each node to the copied set of predecessors/successors as a local variable to the sort function, but if we're talking about storing one set per node, as a field on the node is the more object-oriented way to go about that. Plus, this way the performance cost is a pointer dereference, whereas if it were a map, the cost would include hashing a pointer and then doing equality in a bucket. When decision trees get large, the liveness analysis already takes a long time, so I'd rather keep things fast even if it requires slightly more memory, unless we reach a point where that becomes a problem, which we haven't yet.